### ISSUE TITLE
multiclass_classification prediction score, how is it calculated?

### ISSUE THREAD

**Issue by maxxfan** [Microsoft/LightGBM#285]
May I ask some questions about multiclass prediction process? 

For multiclass_classification example, the prediction result LightGBM_predict_result.txt looks like:
0.35487178523191665     0.27813394980323153     0.11328126210446009     0.059019174521813413    0.19469382833857823
0.092846988782339712    0.13315247488950777     0.23752461867816194     0.2414290772499664      0.29504684040002416
0.13632794556950914     0.21393625584060516     0.29558618898152428     0.2215388013824324      0.13261080822592899
...

Q1: The highest score of each line is the predicted class number, is it right?
Q2: How those score are calculated? More specifically, how are they related to leaf_value and/or internal_value in the tree?

Thanks a lot in advance.

**Comment by Laurae2** [comment]
Q1: I take your example, in bold is the class predicted with the highest probability:

| Class 1 | Class 2 | Class 3 | Class 4 | Class 5 |
| --- | --- | --- | --- | --- |
| **0.35487**... | 0.27813... | 0.11328... | 0.05901... | 0.19469... |
| 0.09284... | 0.13315... | 0.23752... | 0.24142... | **0.29504**... |
| 0.13632... | 0.21393... | **0.29558**... | 0.22153... | 0.13261... |

By default it is your predicted class but you may give different weights which leads to different predicted classes (this is not LightGBM-specific, it can apply to any multiclass problem).

Q2: it is `1 vs All classifier`, therefore you have a model for each `1 vs Other classes`, then you compute the raw score for each class, and apply the Softmax function. If sum is not equal to 1, then rescale so the sum is equal to 1.

**Comment by guolinke** [comment]
@Laurae2 Thanks for the detailed explanation 👍

**Comment by uygnef** [comment]
how are those raw score  calculated for multiclass classification?

say one tree is, and I have 8 classes:
Tree=1
num_leaves=10
num_cat=0
split_feature=4 16 4 23 4 32 12 4 11
split_gain=413.714 336.229 165.663 151.734 55.6006 47.0078 28.5545 28.1881 26.4878
threshold=29.500000000000004 1.0000000180025095e-35 213.00000000000003 570.00000000000011 46.500000000000007 5522.5000000000009 -454.49999999999994 459.50000000000006 233.50000000000003
decision_type=2 2 2 2 2 2 2 2 2
left_child=-1 2 3 5 -3 6 -2 8 -6
right_child=1 4 -4 -5 7 -7 -8 -9 -10
leaf_value=-0.0080359150505640055 0.0021818181729513749 0.090247936798252659 0.011461450241577032 -0.00029484038590479666 0.006804680612607944 0.059500182987276079 0.089715063954614288 -0.0088529275526329308 0.035294118737154476
leaf_count=729 10 22 158 74 303 251 402 191 119
internal_value=0 1.94011 2.95033 3.45997 0.516249 3.8478 4.37952 0.372831 0.741922
internal_count=2259 1530 895 737 635 663 412 613 422
shrinkage=0.02

After prediction, one sample should have a score which is leaf_value.
But I can only get one score, how can I get the other 7 values to calculate final score by Softmax function?

**Comment by guolinke** [comment]
For multi-class, one iteration will produce `num_class` trees, each presents the score of one class.

**Comment by SoufianeDataFan** [comment]
@Laurae2  How to rescale the probabilities when the sum is not equal to 1?
